### PR TITLE
[MNK] Implement initial rotation after DT rework

### PR DESCRIPTION
**Pull request by sammhill**
Full simple implementation of MNK rotation
Full advanced option implementation of MNK rotation
Full AOE simple rotation

Both openers recommended by The Balance are implemented, with a choice.

Removed some options as raised by Kage, but will be gradually added back with modifications for Dawntrail job changes

### CODE REVIEW

**Comment by ZielonyQ** [comment]
its possible u can add Double Lunar and Solar Lunar openers as separated think to chose? Like it was at EW

[Solar Lunar opener](https://files.catbox.moe/y7pe8k.png)
[Double Lunar opener](https://files.catbox.moe/cy23wy.png)

**Comment by sammhill** [comment]
> its possible u can add Double Lunar and Solar Lunar openers as separated think to chose? Like it was at EW
> 
> [Solar Lunar opener](https://files.catbox.moe/y7pe8k.png) [Double Lunar opener](https://files.catbox.moe/cy23wy.png)

Just added! Give it a try :)

**Comment by ZielonyQ** [comment]
> > its possible u can add Double Lunar and Solar Lunar openers as separated think to chose? Like it was at EW
> > [Solar Lunar opener](https://files.catbox.moe/y7pe8k.png) [Double Lunar opener](https://files.catbox.moe/cy23wy.png)
> 
> Just added! Give it a try :)
> 
> Also updated to latest openers shared in The Balance

yep work good tysm! 😄

**Comment by machinaeZER0** [comment]
Hey! New to this plugin in general, and apologies if this is outside the scope of your specific PR - is there a world in which there's an option that simply consolidates the six main GCDs into three buttons that switch between their variants according to the ball gauge?

**Comment by sammhill** [comment]
> Hey! New to this plugin in general, and apologies if this is outside the scope of your specific PR - is there a world in which there's an option that simply consolidates the six main GCDs into three buttons that switch between their variants according to the ball gauge?

Thanks for the suggestion. Added this morning, give it a try :)

**Comment by Scjaro** [comment]
sometimes it trys to "double Rising Phoenix" in the Solar Lunar Opener or "double Elixier Burst" in the Double Lunar opener

example with Solar Lunar Opener:
image

**Comment by machinaeZER0** [comment]
I haven't had a chance to test this (apologies), but in terms of the 6 GCD > 3 GCD button consolidation, this will gracefully handle the lower levels/content syncs where ball spenders aren't yet unlocked, correct? I think that's what the code implied when I was skimming through, but wanted to triple check.

Thanks again for putting the time in here!

**Comment by Warchaild** [comment]
Issues I'm experiencing here -

Looping PC instead of demo when appropriate (opener is fine, happens after the opener.  Will constantly use PC even with no balls)
TFC pushing GCDs (might just need to up the timing)
One minute and two minute windows overwriting the formless after a PB finisher with fires reply (currently behaving like PB > Fire's Reply instead of PB > Opo > Fire's Reply > Opo)

**Comment by Luschel666** [comment]
When is MNK scheduled to be implemented?

**Comment by sammhill** [comment]
> I haven't had a chance to test this (apologies), but in terms of the 6 GCD > 3 GCD button consolidation, this will gracefully handle the lower levels/content syncs where ball spenders aren't yet unlocked, correct? I think that's what the code implied when I was skimming through, but wanted to triple check.
> 
> Thanks again for putting the time in here!

Yep, I believe so.



> Excuse me for the comment. When do you plan to implement MNK?

Whenever this pull request gets merged is when MNK will be available in the main Sloth repo

---

For anyone reading, I've just pushed some changes to better align the buff windows with the rotation. Please give it a try!

**Comment by Cadwgan** [comment]
Few minutes of dummy bashing and not seeing any issues with ST, little drifting but I also don't have a 1.94 GCD currently.

Noticed AoE isn't dumping chakra at 6+, code seems to be checking for exactly 5 chakra rather than 5 or more.

One request would be for Formless Strikes to change Bootshine to DK when outside the opener with no opo fury, mostly for dungeons or extended downtime without a target, currently it only works when all CDs for the opener are ready.

**Comment by Warchaild** [comment]
Having a lot of issues with the opener now.  

Solar/Lunar is trying to use PB again while PB is up in the opener.  It's also blowing RoF and Brotherhood early, and not using Fire's Reply in the right place like it was before.  

Double Lunar does every buff then tries to use Brotherhood again when its already on CD, and also doesn't use Fire's Reply in the right place.  

The PC looping issue from my previous comment is still there as well, using either opener option.

**Comment by Cadwgan** [comment]
Try deselecting and reselecting all the options you're using, also drag bootshine back to the bar, It was spamming the 2nd PB until it escaped the opener for 2 tries for me after updating. If you're still having issues post what options you have enabled so others can try to replicate it.

**Comment by AstragothIronhand** [comment]
> Try deselecting and reselecting all the options you're using, also drag bootshine (not opo) back to the bar, It was spamming the 2nd PB until it escaped the opener for 2 tries for me after updating. If you're still having issues post what options you have enabled so others can try to replicate it.

Yep, tried that, reinstanced, tried the different openers etc.  

I have everything selected except the combo heals option

**Comment by machinaeZER0** [comment]
Posted on Discord, but reposting here in case others are able to corroborate/help test: I'm testing out the "ball handlers" options for Monk (options 18-21), where the six single target form attacks are consolidated into three buttons. 19 and 20 seem to be working correctly for me so far (opo opo + raptor), but 21 (coerl) is not swapping between snap punch and demolish for me.

more context:
- I currently have both snap punch and demolish on my hotbar (freshly dragged from Actions and Traits), and neither one changes as i go through my rotation
- I am currently level 81
- 18-21 are the only MNK options i have checked currently
- I disabled the release version of the plugin before enabling this one
- I have only tried this on a training dummy so far

My current hypothesis is that the coerl patch may be reversed - that it thinks snap punch is a ball filler and not a ball spender, and vice versa? If so then it would make sense that it would never pass the check to swap attacks.

===

Additional semi-related issues I'm seeing with the above options toggled on:
- popping perfect balance is not letting me alternate opos - the button only changes form a single time upon execution of the skill
- I'm not 100% sure formless fist is working correctly if i disengage mid-combat -  i would expect all three forms to switch to their "lit up"/available variants in that situation, but that's not currently the case from what I can see. It does seem to be working correctly pre-pull, but I think there's a chance it's coded specifically to handle an opener situation?

**Comment by sammhill** [comment]
> Posted on Discord, but reposting here in case others are able to corroborate/help test: I'm testing out the "ball handlers" options for Monk (options 18-21), where the six single target form attacks are consolidated into three buttons. 19 and 20 seem to be working correctly for me so far (opo opo + raptor), but 21 (coerl) is not swapping between snap punch and demolish for me.
> 
> more context:
> 
> * I currently have both snap punch and demolish on my hotbar (freshly dragged from Actions and Traits), and neither one changes as i go through my rotation
> * I am currently level 81
> * 18-21 are the only MNK options i have checked currently
> * I disabled the release version of the plugin before enabling this one
> * I have only tried this on a training dummy so far
> 
> My current hypothesis is that the coerl patch may be reversed - that it thinks snap punch is a ball filler and not a ball spender, and vice versa? If so then it would make sense that it would never pass the check to swap attacks.
> 
> ===
> 
> Additional semi-related issues I'm seeing with the above options toggled on:
> 
> * popping perfect balance is not letting me alternate opos - the button only changes form a single time upon execution of the skill
> * I'm not 100% sure formless fist is working correctly if i disengage mid-combat -  i would expect all three forms to switch to their "lit up"/available variants in that situation, but that's not currently the case from what I can see. It does seem to be working correctly pre-pull, but I think there's a chance it's coded specifically to handle an opener situation?

Have just pushed some fixes for what you've stated. I'm however not getting the Snap Punch/Demolish issue, the ball handler works on snap punch so that should be the only one you need on your bar

**Comment by machinaeZER0** [comment]
> Have just pushed some fixes for what you've stated. I'm however not getting the Snap Punch/Demolish issue, the ball handler works on snap punch so that should be the only one you need on your bar

Thank you! So odd about Snap Punch... building the updated fork now, will try to test later with a fresh install. Are settings saved locally somewhere that I could/should flush?

**Comment by machinaeZER0** [comment]
The fixes to perfect balance and formless fist seem to both be working great with the latest build, thank you! Even after doing a "reset plugin data and reload" though, snap punch and demolish (freshly pulled to my bar again for testing) are still not swapping with each other with 21 enabled :( so odd...

Which parts of the PR handle option 21 specifically? I can try and look over the code itself later, but if you're not experiencing it then that's quite puzzling. Any chance it has to do with my level (still 81)/not having pouncing coerl available?

Is anyone else able to test this too? ^ would be wild if it were just me, lol

**Comment by sammhill** [comment]
Have just pushed a few fixes, but have realised there are a few things still broken in this, will be fixing hopefully tomorrow

**Comment by AstragothIronhand** [comment]
> Issues I'm experiencing here -
> 
> Looping PC instead of demo when appropriate (opener is fine, happens after the opener. Will constantly use PC even with no balls) TFC pushing GCDs (might just need to up the timing) One minute and two minute windows overwriting the formless after a PB finisher with fires reply (currently behaving like PB > Fire's Reply instead of PB > Opo > Fire's Reply > Opo)

Today's fixes fixed the opener and one minute/two minute windows for me.  Excellent work!

I'm still noticing that TFC is pushing GCDs, and the coeurl GCDs are still not replacing correctly in the stack with no balls.

**Comment by machinaeZER0** [comment]
Found another bug that may be related to my snap punch/demo issue - just ran a leveling roulette and got thousand maws of toto-rak, and while I was level synced the plugin was still trying to alternate opo GCDs even though dragon kick wasn't available. Also, sometimes snap punch wouldn't alternate to the dragon kick icon, but it also wasn't lit up and wouldn't let me press it, so I had to hit Arm of the Destroyer instead to keep my ST combo going, hahah.

Might be unrelated to my previous issue, but it does seem like options 17-20 may currently have issues in general when not at max level. Hope this is helpful info, let me know if I can be of further assistance!

**Comment by Cadwgan** [comment]
[XIVanalasys link](https://xivanalysis.com/fflogs/GcbjpmWfY6CVDBJH/8/1)

Opener issue, TFC being used at certain points makes the opener stall, in this case right after Fire's Reply, it kept trying to use the GCD again for a few seconds even though it was spent.

I'm unable to replicate this on a dummy, but have had this happen a few times in group content.

**Comment by Cadwgan** [comment]
image

Similar problem, this time on Elixir Burst.

It seems to be the usage of TFC is sometimes preventing the opener from proceeding a step.

**Comment by AstragothIronhand** [comment]
> > Issues I'm experiencing here -
> > Looping PC instead of demo when appropriate (opener is fine, happens after the opener. Will constantly use PC even with no balls) TFC pushing GCDs (might just need to up the timing) One minute and two minute windows overwriting the formless after a PB finisher with fires reply (currently behaving like PB > Fire's Reply instead of PB > Opo > Fire's Reply > Opo)
> 
> Today's fixes fixed the opener and one minute/two minute windows for me. Excellent work!
> 
> I'm still noticing that TFC is pushing GCDs, and the coeurl GCDs are still not replacing correctly in the stack with no balls.

Demo and PC are changing as expected now.  1 minute and 2 minute windows look good.  I have no further issues.

**Comment by AstragothIronhand** [comment]
Just ran through m2s and noticed that only about half of my fires reply are being used after an opo action.  Any way to add a check here to force it to behave as such?

**Review by Taurenkey — CHANGES_REQUESTED** [review]
Apologies for taking so long to get around to this. This is all over the place and not structured properly.

- Labelling issues:
- - The correct terminology we use is "Simple Mode" for one button rotations without customization options. In this case, it should be `Simple Mode - AoE` rather than just `AoE Rotation`. Also the description should be updated to reflect other simple modes that have been converted to this style.
- - Likewise for ST, it should be `Advanced Mode - Single Target` with the appropriate description template used. 
- - `Balls Handler`? It's not really appropriate, the correct terminology for these mechanics are "Beast Chakras"
- - Sub-option labels should where possible follow the template `X Option`, e.g `Meditation Option`, with the description `Adds X to the rotation.` e.g `Adds Meditation to the rotation.`
- Missing Simple Mode
- - There's a Simple Mode for AoE but not for ST? The very least we require a job to have is simple modes for both ST & AoE, advanced modes can come later. Since there's an advanced mode, you'll just need to copy it but remove all custom user options and give them predefined settings instead.

**Comment by sammhill** [comment]
@Taurenkey - Have made the changes you've recommended here, apologies on `Ball Handlers`, it was just the common name for the process from the Balance so thought it would make sense here.

**Comment by sammhill** [comment]
@machinaeZER0 - Good suggestion, pushed! :)

File: XIVSlothCombo/Combos/CustomComboPreset.cs [hunk 1]
```diff
@@ -1834,136 +1834,103 @@ public enum CustomComboPreset
...
 
-        [ParentCombo(MNK_ST_SimpleMode)]
-        [CustomComboInfo("Dynamic True North Option", "Adds True North to the main combo right before positionals if you aren't in the correct position for their bonuses.", MNK.JobID)]
-        MNK_TrueNorthDynamic = 9029,
+        [ReplaceSkill(MNK.TrueStrike)]
+        [CustomComboInfo("Raptor Option", "Replace True Strike/Rising Raptor with Twin Snakes.", MNK.JobID)]
+        [ParentCombo(MNK_ST_BeastChakras)]
+        MNK_BC_RAPTOR = 9021,
 
-        [Variant]
```
Comment (Tartarga) [inline XIVSlothCombo/Combos/CustomComboPreset.cs:1955]: Please restore
This is for the Variant Dungeons

File: XIVSlothCombo/Combos/CustomComboPreset.cs [hunk 2]
```diff
@@ -1834,136 +1834,103 @@ public enum CustomComboPreset
...
-        [VariantParent(MNK_ST_SimpleMode, MNK_AoE_SimpleMode)]
-        [CustomComboInfo("Cure Option", "Use Variant Cure when HP is below set threshold.", MNK.JobID)]
-        MNK_Variant_Cure = 9030,
+        [ReplaceSkill(MNK.SnapPunch)]
+        [CustomComboInfo("Coeurl Option", "Replace Snap Punch/Pouncing Coeurl with Demolish.", MNK.JobID)]
+        [ParentCombo(MNK_ST_BeastChakras)]
+        MNK_BC_COEURL = 9022,
 
-        [Variant]
```
Comment (Tartarga) [inline XIVSlothCombo/Combos/CustomComboPreset.cs:1960]: Please restore
This is for the Variant Dungeons

File: XIVSlothCombo/Window/Functions/UserConfig.cs [hunk 3]
```diff
@@ -1615,27 +1615,18 @@ internal static void Draw(CustomComboPreset preset, bool enabled)
...
-            if (preset == CustomComboPreset.MNK_AoE_ComboHeals)
+            if (preset == CustomComboPreset.MNK_STUseOpener && enabled)
             {
-                UserConfig.DrawSliderInt(0, 100, MNK.Config.MNK_AoESecondWindThreshold, "Second Wind HP percentage threshold (0 = Disabled)", 150, SliderIncrements.Ones);
-                UserConfig.DrawSliderInt(0, 100, MNK.Config.MNK_AoEBloodbathThreshold, "Bloodbath HP percentage threshold (0 = Disabled)", 150, SliderIncrements.Ones);
+                UserConfig.DrawHorizontalRadioButton(MNK.Config.MNK_SelectedOpener, "Double Lunar", "Uses Lunar/Lunar opener", 1);
+                UserConfig.DrawHorizontalRadioButton(MNK.Config.MNK_SelectedOpener, "Solar Lunar", "Uses Solar/Lunar opener", 2);
             }
 
-            if (preset == CustomComboPreset.MNK_Variant_Cure)
```
Comment (Tartarga) [inline XIVSlothCombo/Window/Functions/UserConfig.cs:1636]: Please restore
This is for the Variant Dungeons

File: XIVSlothCombo/Combos/PvE/MNK.cs [hunk 4]
```diff
@@ -75,639 +88,608 @@ public const byte
...
-                            return Thunderclap;
+                            return OriginalHook(Meditation); ;
                         }
                     }
 
                     // Buffs
-                    if (inCombat && !inOpener)
+                    if (inCombat && canWeave)
                     {
-                        if (IsEnabled(CustomComboPreset.MNK_Variant_Rampart) &&
```
Comment (Tartarga) [inline XIVSlothCombo/Combos/PvE/MNK.cs:485]: Restore the Preset Check. Not all users want this on in Variant Dungeons.

File: XIVSlothCombo/Combos/PvE/MNK.cs [hunk 5]
```diff
@@ -75,639 +88,608 @@ public const byte
...
-                        }
-
-                        if (IsEnabled(CustomComboPreset.MNK_AoE_Simple_Thunderclap) && !InMeleeRange() && gauge.Chakra == 5 && (!LevelChecked(FormShift) || HasEffect(Buffs.FormlessFist)))
-                        {
-                            return Thunderclap;
-                        }
+                        return OriginalHook(Meditation);
                     }
 
-                    if (IsEnabled(CustomComboPreset.MNK_Variant_Cure) && IsEnabled(Variant.VariantCure) && PlayerHealthPercentageHp() <= GetOptionValue(Config.MNK_VariantCure))
```
Comment (Tartarga) [inline XIVSlothCombo/Combos/PvE/MNK.cs:133]: Please Restore for Variant Dungeons

File: XIVSlothCombo/Combos/PvE/MNK.cs [hunk 6]
```diff
@@ -75,639 +88,608 @@ public const byte
...
-                        return Variant.VariantCure;
-
-                    // Buffs
+                    // OGCDs
                     if (inCombat && canWeave)
                     {
-                        if (IsEnabled(CustomComboPreset.MNK_Variant_Rampart) &&
```
Comment (Tartarga) [inline XIVSlothCombo/Combos/PvE/MNK.cs:139]: Please Restore for Variant Dungeons

**Review by Tartarga — CHANGES_REQUESTED** [review]
Infrastructure Variant Dungeons actions were disabled or removed.  Please restore this functionality. I marked what I saw but I could've missed something

**Comment by Jamein136** [comment]
Works great for the most part. Main thing I have noticed is masterful blitz will just not be pressed if you are outside of riddle of fire. 
If for example your 2nd PB during 2 minutes ends up finishing late for downtime reasons and RoF drops off the combo will just continue on as normal but ignore masterful blitz finisher and leave it there unless you press manually.
Not a huge deal to fix yourself as you're playing but just something I noticed. 

Generally happens if the combo misses starting Perfect balance when RoF has under 4 seconds on its cooldown

**Comment by sammhill** [comment]
> Infrastructure Variant Dungeons actions were disabled or removed. Please restore this functionality. I marked what I saw but I could've missed something

Have re-added these as requested

File: XIVSlothCombo/Window/Functions/UserConfig.cs [hunk 7]
```diff
@@ -1615,22 +1615,16 @@ internal static void Draw(CustomComboPreset preset, bool enabled)
...
-                UserConfig.DrawSliderInt(0, 100, MNK.Config.MNK_STBloodbathThreshold, "Bloodbath HP percentage threshold (0 = Disabled)", 150, SliderIncrements.Ones);
+                UserConfig.DrawSliderInt(0, 100, MNK.Config.MNK_ST_SecondWind_Threshold, "Second Wind HP percentage threshold (0 = Disabled)", 150, SliderIncrements.Ones);
+                UserConfig.DrawSliderInt(0, 100, MNK.Config.MNK_ST_Bloodbath_Threshold, "Bloodbath HP percentage threshold (0 = Disabled)", 150, SliderIncrements.Ones);
             }
 
-            if (preset == CustomComboPreset.MNK_AoE_ComboHeals)
+            if (preset == CustomComboPreset.MNK_STUseOpener && enabled)
             {
-                UserConfig.DrawSliderInt(0, 100, MNK.Config.MNK_AoESecondWindThreshold, "Second Wind HP percentage threshold (0 = Disabled)", 150, SliderIncrements.Ones);
-                UserConfig.DrawSliderInt(0, 100, MNK.Config.MNK_AoEBloodbathThreshold, "Bloodbath HP percentage threshold (0 = Disabled)", 150, SliderIncrements.Ones);
```
Comment (zbee) [inline XIVSlothCombo/Window/Functions/UserConfig.cs:1633]: Looks like AoE healing options are missing, despite still being referenced.
https://github.com/Nik-Potokar/XIVSlothCombo/blob/72071cd96b5957274381a4883e3a7158a90b1ce0/XIVSlothCombo/Combos/PvE/MNK.cs#L111

Comment (sammhill) [reply XIVSlothCombo/Window/Functions/UserConfig.cs:1633]: Removed this config, will re-implement at a later date when AoE advanced is added

**Comment by TimeBread** [comment]
Slight optimization note.  On every 2 minute burst after the opener (using double lunar option), the combo will prioritize using the Rising Phoenix combo (standard combo) instead of the more powerful Elixir Burst combo (opo spam) (at level 100).

**Comment by Cadwgan** [comment]
@TimeBread Honestly, this entirely depends on your fury at the time and triple opo being more powerful is pre-7.05 information.

The average 123, 111, 222 or 333 for MNK now is within 10 potency of each other, with opo falling behind with more crit, unless we want to start coding for each permutation of fury going into burst windows it's probably best left as it is.

**Pull request merged**